### PR TITLE
Jwt library

### PR DESCRIPTION
**Pull request by ChTimTsubasa**
Implemented JWT library.

### CODE REVIEW

**Comment by smtakeda** [comment]
Fix Mac and Windows tests, for example:
https://travis-ci.org/snowflakedb/libsnowflakeclient/jobs/426598146
```
/Users/travis/build/snowflakedb/libsnowflakeclient/cpp/jwt/Signer.cpp:84:73: error: if statement has empty body [-Werror,-Wempty-body]
  if (1 != EVP_DigestSignUpdate(mdctx.get(), msg.c_str(), msg.length()));
                                                                        ^
```

File: cpp/jwt/Signer.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+{
+namespace Jwt
+{
+std::map<std::string, AlgorithmType>
+  AlgorithmTypeMapper::reverse_map_ = {
+  {HS_256,  AlgorithmType::HS256},
+  {HS_384,  AlgorithmType::HS384},
+  {HS_512,  AlgorithmType::HS512},
+  {RS_256,  AlgorithmType::RS256},
+  {RS_256,  AlgorithmType::RS256},
```
Comment (howryu) [inline cpp/jwt/Signer.cpp]: Why two RS_256?

File: cpp/jwt/Util.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2017-2018 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include "Util.hpp"
+#include "JwtException.hpp"
+#include "../util/Base64.hpp"
```
Comment (howryu) [inline cpp/jwt/Util.cpp]: Use absolute path include.

File: cpp/jwt/Util.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  std::vector<char> decoded(decode_len);
+  decode_len = Client::Util::Base64::decodeUrl(in_text.c_str(), in_text.length(), decoded.data());
+  if ((ssize_t) decode_len == -1) throw JwtParseFailure();
+
+  decoded.resize(decode_len);
+  return decoded;
+}
+
+}
+}
```
Comment (howryu) [inline cpp/jwt/Util.cpp]: newline at the end

**Comment by smtakeda** [comment]
Windows build error:
```
c:\projects\libsnowflakeclient\cpp\jwt\ClaimSet.hpp(145): error C2039: 'function': is not a member of 'std' [C:\projects\libsnowflakeclient\cmake-build-win64\snowflakeclient.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\vector(17): note: see declaration of 'std'
c:\projects\libsnowflakeclient\cpp\jwt\ClaimSet.hpp(145): error C2065: 'function': undeclared identifier [C:\projects\libsnowflakeclient\cmake-build-win64\snowflakeclient.vcxproj]
c:\projects\libsnowflakeclient\cpp\jwt\ClaimSet.hpp(145): error C2059: syntax error: ')' [C:\projects\libsnowflakeclient\cmake-build-win64\snowflakeclient.vcxproj]
c:\projects\libsnowflakeclient\cpp\jwt\ClaimSet.hpp(145): error C2059: syntax error: '>' [C:\projects\libsnowflakeclient\cmake-build-win64\snowflakeclient.vcxproj]
c:\projects\libsnowflakeclient\cpp\jwt\ClaimSet.hpp(145): error C2238: unexpected token(s) preceding ';' [C:\projects\libsnowflakeclient\cmake-build-win64\snowflakeclient.vcxproj]
...
```

File: cpp/jwt/Util.cpp (hunk 3 above)
Comment (ChTimTsubasa) [reply cpp/jwt/Util.cpp]: Addressed

**Comment by notkriswagner** [comment]
LGTM. Merging

**Pull request merged**